### PR TITLE
Benchmark: Initialize spheres on demand 

### PR DESCRIPTION
**Pull request by elviscapiaq**
- It considerably reduces the startup time and memory by initializing only 100 spheres.

### CODE REVIEW

File: src/ppx/knob.cpp [hunk 1]
```diff
@@ -49,7 +49,6 @@ KnobCheckbox::KnobCheckbox(const std::string& flagName, bool defaultValue)
     : Knob(flagName, true), mValue(defaultValue), mDefaultValue(defaultValue)
 {
     SetFlagParameters("<true|false>");
-    RaiseUpdatedFlag();
```
Comment (angela28chen) [inline src/ppx/knob.cpp:52]: Removing `RaiseUpdatedFlag()` from different knob classes is going to affect the behaviour of all knobs. I don't think this PR adequately addresses that.

Comment (elviscapiaq) [reply src/ppx/knob.cpp]: Thanks for pointing it. I restored the calls to `RaiseUpdatedFlag()` and updated the PR.

**Review by wangra-google — APPROVED** [review]

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp [hunk 2]
```diff
@@ -694,28 +700,34 @@ void GraphicsBenchmarkApp::ProcessKnobs()
...
+    const bool enableSpheresKnobChanged      = pEnableSpheres->DigestUpdate();
+    const bool initializedSpheresKnobChanged = pInitializedSpheres->DigestUpdate();
 
     const bool enableSpheres = pEnableSpheres->GetValue();
     // If the LOD is empty, assuming we skipped the setup for all sphere resources at start
     // So need to do the intial setup for all resources here
     const bool spheresAreSetUp               = !mSphereLODs.empty();
     const bool updateSphereDescriptors       = spheresAreSetUp && allTexturesTo1x1KnobChanged;
-    const bool setupSphereResourcesAndMeshes = (!spheresAreSetUp) && enableSpheres;
+    const bool setupSphereResourcesAndMeshes = (!spheresAreSetUp || initializedSpheresKnobChanged) && enableSpheres;
```
Comment (Keenuts) [inline benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:714]: Sphere resources are not cleaned up. Right now, we re-create new descriptors while keeping the old ones, exhausting the pool. This needs to be fixed before this knob can be used.

The other issue is we initialize spheres twice by default:
 - once because the pEnableSpheres bit is set by default.
 - then here, because the initializedSpheresKnob by default is marked as changed.

**Review by Keenuts — CHANGES_REQUESTED** [review]

**Comment by hysw** [comment]
I don't think we need to have a separate knob for it
i.e. [eeb387c9](https://github.com/google/bigwheels/commit/eeb387c93a75b31e62adcfe3fe773a3d27d966d6)

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp (hunk 2 above)
Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: Updated.

**Comment by elviscapiaq** [comment]
Thank you for all your feedback. I have modified the essence of this PR and implemented on-demand initialization for the spheres. Please let me know your thoughts.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp [hunk 3]
```diff
@@ -544,19 +535,37 @@ void GraphicsBenchmarkApp::SetupSkyBoxMeshes()
...
 
     // Create the meshes
     uint32_t meshIndex = 0;
     for (const auto& lod : kAvailableLODs) {
         PPX_LOG_INFO("LOD: " << lod.name);
         SphereMesh sphereMesh(/* radius = */ 1, lod.value.longitudeSegments, lod.value.latitudeSegments);
         sphereMesh.ApplyGrid(grid);
-
         // Create a giant vertex buffer for each vb type to accommodate all copies of the sphere mesh
+        if (mSphereMeshes[meshIndex] != nullptr) {
```
Comment (Keenuts) [inline benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:551]: Is there a case where not all meshes are initialized?
I think it would be similar to just add this before the loop:

```
for (auto& mesh : mSphereMeshes)
    GetDevice()->DestroyMesh(mesh);
```

Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: I think at the beginning all meshes are nullptr,. Updated.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp [hunk 4]
```diff
@@ -544,15 +535,25 @@ void GraphicsBenchmarkApp::SetupSkyBoxMeshes()
...
+    const uint32_t requiredSphereCount = std::max<uint32_t>(pSphereInstanceCount->GetValue(), kDefaultSphereInstanceCount);
+    const uint32_t initSphereCount     = std::min<uint32_t>(kMaxSphereInstanceCount, 2 * std::max(requiredSphereCount, mInitializedSpheres));
+
+    OrderedGrid grid(initSphereCount, kSeed);
+    mInitializedSpheres = initSphereCount;
+
+    // Destroy the meshes if they were created.
+    for (auto& mesh : mSphereMeshes) {
+        if (mesh != nullptr) {
+            GetDevice()->DestroyMesh(mesh);
```
Comment (hysw) [inline benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:547]: nit: reset it back to nullptr after you destroyed it, try not to leave dangling pointer around.

Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:547]: Done.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp [hunk 5]
```diff
@@ -874,13 +876,14 @@ void GraphicsBenchmarkApp::ProcessKnobs()
...
     const bool enableSpheresKnobChanged    = pEnableSpheres->DigestUpdate();
+    const bool requireMoreSpheres          = (pSphereInstanceCount->GetValue() > mInitializedSpheres);
 
     const bool enableSpheres = pEnableSpheres->GetValue();
     // If the LOD is empty, assuming we skipped the setup for all sphere resources at start
     // So need to do the intial setup for all resources here
     const bool spheresAreSetUp               = mSpheresAreSetUp;
     const bool updateSphereDescriptors       = spheresAreSetUp && allTexturesTo1x1KnobChanged;
-    const bool setupSphereResourcesAndMeshes = (!spheresAreSetUp) && enableSpheres;
+    const bool setupSphereResourcesAndMeshes = (!spheresAreSetUp || requireMoreSpheres) && enableSpheres;
```
Comment (angela28chen) [inline benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:886]: I believe this change will cause `SetupSphereResources()` to be called more than once, but that has logic like loading shaders which probably should only be called once.

Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: Update.

**Comment by elviscapiaq** [comment]
> Note that because of the way imGUI sliders work, the value will be adjusted as the slider is dragged and not when it's let go. So with this change you will need to drag the slider ~4 times upwards and wait while it buffers to slowly make it to 3000, it's a little frustrating to use. Do we want to change it so that the # of spheres only change after the slider is let go? That can probably be done if a separate variable is kept.

Thanks for this feedback. I fixed this issue and now the spheres are initialized only when the slider is let go.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp [hunk 6]
```diff
@@ -544,15 +535,26 @@ void GraphicsBenchmarkApp::SetupSkyBoxMeshes()
...
 void GraphicsBenchmarkApp::SetupSphereMeshes()
 {
-    OrderedGrid grid(kMaxSphereInstanceCount, kSeed);
+    const uint32_t requiredSphereCount = std::max<uint32_t>(pSphereInstanceCount->GetValue(), kDefaultSphereInstanceCount);
+    const uint32_t initSphereCount     = std::min<uint32_t>(kMaxSphereInstanceCount, 2 * std::max(requiredSphereCount, mInitializedSpheres));
+
+    OrderedGrid grid(initSphereCount, kSeed);
+    mInitializedSpheres = initSphereCount;
+
+    // Destroy the meshes if they were created.
```
Comment (Keenuts) [inline benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:544]: This triggers a Vulkan validation error:
The destroy can be called while the buffer is still used by the previous frame.
We might want to do a WaitForIdle (quick solution, don't need perf there) or something more fine grained to prevent those issues.

Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:545]: Done.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp [hunk 7]
```diff
@@ -1593,9 +1607,9 @@ void GraphicsBenchmarkApp::RecordCommandBufferSpheres(PerFrame& frame)
     frame.cmd->BindGraphicsDescriptorSets(mSphere.pipelineInterface, 1, &mSphere.descriptorSets.at(GetInFlightFrameIndex()));
 
     // Snapshot some scene-related values for the current frame
-    uint32_t currentSphereCount   = pSphereInstanceCount->GetValue();
+    uint32_t currentSphereCount   = std::min<uint32_t>(pSphereInstanceCount->GetValue(), mInitializedSpheres);
```
Comment (wangra-google) [inline benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:1610]: in what case `pSphereInstanceCount` could be larger than `mInitializedSpheres`?

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp [hunk 8]
```diff
@@ -1593,9 +1607,9 @@ void GraphicsBenchmarkApp::RecordCommandBufferSpheres(PerFrame& frame)
...
     uint32_t currentDrawCallCount = pDrawCallCount->GetValue();
-    uint32_t mSphereIndexCount    = mSphereMeshes[meshIndex]->GetIndexCount() / kMaxSphereInstanceCount;
+    uint32_t mSphereIndexCount    = (mInitializedSpheres > 0 ? (mSphereMeshes[meshIndex]->GetIndexCount() / mInitializedSpheres) : 0);
```
Comment (wangra-google) [inline benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:1612]: in what case `mInitializedSpheres` could be 0? If i understand correctly, only when `SetupSphereMeshes` is skipped,  mInitializedSpheres is 0? but in that case, `pEnableSpheres` is false, and  `RecordCommandBufferSpheres` would be skipped as well?

Comment (Keenuts) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:1612]: I don't think those case could be reached as-is. But since those values depends on elements external to this function, I would be in favor of keeping this check and min.
Just so we say "in case this happens, we have a defined behavior to handle it", even if if wouldn't happen today.

**Review by Keenuts — APPROVED** [review]

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp (hunk 8 above)
Comment (wangra-google) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:1612]: if `mInitializedSpheres` should never be 0, maybe better with an assert?

**Review by angela28chen — APPROVED** [review]
Not something that needs to change for this PR, but I think that there might be a need for two different types of sliders for slider knobs, one where the value is reflected immediately during dragging and another where it's only updated once the slider is released. The current slider knob logic is not really consistent with itself. I think that `pSphereInstanceCount` shows this because `RecordCommandBufferSpheres()` will use the value (which is changing as the knob is getting dragged) whereas the vertex buffer will be using the value only after it has been marked as updated (which is when the knob is released).

So this is probably fine for this PR if dragging around the sphere slider isn't causing any index out of bounds crashes, but it's something that should be changed in knob implementation.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp (hunk 8 above)
Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: Updated.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp [hunk 9]
```diff
@@ -874,14 +877,16 @@ void GraphicsBenchmarkApp::ProcessKnobs()
...
     // If the LOD is empty, assuming we skipped the setup for all sphere resources at start
     // So need to do the intial setup for all resources here
-    const bool spheresAreSetUp               = mSpheresAreSetUp;
-    const bool updateSphereDescriptors       = spheresAreSetUp && allTexturesTo1x1KnobChanged;
-    const bool setupSphereResourcesAndMeshes = (!spheresAreSetUp) && enableSpheres;
-    const bool rebuildSpherePipeline         = setupSphereResourcesAndMeshes || (spheresAreSetUp && (alphaBlendKnobChanged || depthTestWriteKnobChanged));
+    const bool spheresAreSetUp         = mSpheresAreSetUp;
+    const bool updateSphereDescriptors = spheresAreSetUp && allTexturesTo1x1KnobChanged;
+    const bool setupSphereResources    = (!spheresAreSetUp) && enableSpheres;
+    const bool setupSpheredMeshes      = ((!spheresAreSetUp) || requireMoreSpheres) && enableSpheres;
```
[the remark this answers is not in the document]
Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: Done.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp [hunk 10]
```diff
@@ -874,14 +877,16 @@ void GraphicsBenchmarkApp::ProcessKnobs()
     const bool alphaBlendKnobChanged       = pAlphaBlend->DigestUpdate();
     const bool depthTestWriteKnobChanged   = pDepthTestWrite->DigestUpdate();
     const bool enableSpheresKnobChanged    = pEnableSpheres->DigestUpdate();
+    const bool requireMoreSpheres          = pSphereInstanceCount->DigestUpdate() && (pSphereInstanceCount->GetValue() > mInitializedSpheres);
```
[the remark this answers is not in the document]
Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: Done.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp (hunk 7 above)
Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:1611]: It can occur when we drag the `sphereCount` slider, because we only update the `initializedSpheres` value once the slider is released. In other words, we can drag the slider, and `sphereCount` can increase, but `initializedSpheres` would remain the same up to the slider is let go.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp (hunk 8 above)
Comment (Keenuts) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: Yes, an assert is also fine if that's an illegal behavior.
But I'd feel like having the slider default to 0 is a valid behavior.

Comment (wangra-google) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: the min value of pSphereInstanceCount is set to 1, so i guess 0 is not valid.

Comment (wangra-google) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: actually, maybe it is better to skip `RecordCommandBufferSpheres()` entirely if `mInitializedSpheres` is 0?

Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: I agree with that change. Updated.

**Comment by elviscapiaq** [comment]
PR was rebased.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp [hunk 11]
```diff
@@ -872,18 +875,21 @@ void GraphicsBenchmarkApp::ProcessKnobs()
...
-    const bool spheresAreSetUp               = mSpheresAreSetUp;
-    const bool updateSphereDescriptors       = spheresAreSetUp && allTexturesTo1x1KnobChanged;
-    const bool setupSphereResourcesAndMeshes = (!spheresAreSetUp) && enableSpheres;
-    const bool rebuildSpherePipeline         = setupSphereResourcesAndMeshes || (spheresAreSetUp && (alphaBlendKnobChanged || depthTestWriteKnobChanged));
+    const bool spheresAreSetUp         = mSpheresAreSetUp;
+    const bool requireMoreSpheres      = sphereInstanceCountKnobChanged && (pSphereInstanceCount->GetValue() > mInitializedSpheres);
+    const bool updateSphereDescriptors = spheresAreSetUp && allTexturesTo1x1KnobChanged;
+    const bool setupSphereResources    = (!spheresAreSetUp) && enableSpheres;
+    const bool setupSphereMeshes       = ((!spheresAreSetUp) || requireMoreSpheres) && enableSpheres;
+    const bool rebuildSpherePipeline   = setupSphereMeshes || (spheresAreSetUp && (alphaBlendKnobChanged || depthTestWriteKnobChanged));
```
Comment (apazylbe) [inline benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:892]: Do we need to rebuild pipelines if there was an increase in the number of spheres but nothing else changed?

Comment (apazylbe) [inline benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:892]: This is a bit hard to follow, and used later in the code where you have to scroll up to figure out when something is rebuilt. Also it's a little weird to update `mSpheresAreSetUp` in `SetupSphereMeshes()`, since mSpheresAreSetUp basically means everything is setup instead of just spheres. So if I call just `SetupSpheresMeshes()` mSpheresAreSetUp will be true. 
Maybe not for this PR, but we should  probably consider cleaning up the logic a bit, one alternative: https://github.com/google/bigwheels/commit/9e5c0b8319a8f6503d3631cee46285aab3c68fa5

Comment (apazylbe) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:892]: Not directly related to this PR, but I wonder if it should also take into account the other knob changes (vertex buffer format, layout, and shaders) or if it can be ommitted altogether since `GetSpherePipeline()` in `RecordCommandBufferSpheres` will create the necessary pipeline anyway

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp [hunk 12]
```diff
@@ -90,7 +90,7 @@ void GraphicsBenchmarkApp::InitKnobs()
     pSphereInstanceCount->SetFlagDescription("Select the number of spheres to draw on the screen.");
     pSphereInstanceCount->SetIndent(1);
 
-    GetKnobManager().InitKnob(&pDrawCallCount, "drawcall-count", /* defaultValue = */ 1, /* minValue = */ 1, kMaxSphereInstanceCount);
+    GetKnobManager().InitKnob(&pDrawCallCount, "drawcall-count", /* defaultValue = */ kDefaultSphereInstanceCount, /* minValue = */ 1, kMaxSphereInstanceCount);
```
[the remark this answers is not in the document]
Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: Good catch. Actually, the default value for pSphereInstanceCount shoud be kDefaultSphereInstanceCount but for drawcall should be 1. Updated.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp [hunk 13]
```diff
@@ -546,15 +537,27 @@ void GraphicsBenchmarkApp::SetupSkyBoxMeshes()
 
 void GraphicsBenchmarkApp::SetupSphereMeshes()
 {
-    OrderedGrid grid(kMaxSphereInstanceCount, kSeed);
+    const uint32_t requiredSphereCount = std::max<uint32_t>(pSphereInstanceCount->GetValue(), kDefaultSphereInstanceCount);
+    const uint32_t initSphereCount     = std::min<uint32_t>(kMaxSphereInstanceCount, 2 * std::max(requiredSphereCount, mInitializedSpheres));
+
+    OrderedGrid grid(initSphereCount, kSeed);
```
[the remark this answers is not in the document]
Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: Done.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp (hunk 11 above)
Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: That is a good point, actually we don't need to rebuild pipelines. so I applied your proposed clean up from the other comment.

Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: Thank you for the comment. I applied your suggested clean up and the code looks more readable now.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp [hunk 14]
```diff
@@ -906,20 +908,27 @@ void GraphicsBenchmarkApp::ProcessKnobs()
...
-        SetupSpheresPipelines();
-    }
-
-    // Update descriptors
-    if (updateSphereDescriptors) {
-        UpdateSphereDescriptors();
+    if (enableSpheres) {
+        // Update sphere resources and mesh
+        const bool spheresAreSetUp = mSpheresAreSetUp;
+        if (!spheresAreSetUp) {
```
Comment (apazylbe) [inline benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:914]: I think we can actually inline now and just check !mSpheresAreSetUp here

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp [hunk 15]
```diff
@@ -906,20 +908,27 @@ void GraphicsBenchmarkApp::ProcessKnobs()
...
+            // This creates all resources.
+            SetupSpheres();
+        }
+        else {
+            const bool requireMoreSpheres = sphereInstanceCountKnobChanged && (pSphereInstanceCount->GetValue() > mInitializedSpheres);
+            if (requireMoreSpheres) {
+                SetupSphereMeshes();
+            }
+            // Rebuild pipelines
+            if (alphaBlendKnobChanged || depthTestWriteKnobChanged) {
```
Comment (apazylbe) [inline benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:924]: Sphere pipeline seems to be dependent on some other knobs like PS, VS, vertex buffer: https://github.com/google/bigwheels/blob/4c9fbb05e77d384456594e6b9ea165b7d0bbcca2/benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp#L738-L747
Do we need to add those here, or perhaps just remove this since I think when we record a command buffer it will compile the necessary pipeline? 
@hysw, @angela28chen

Comment (hysw) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:924]: No, this part can be removed.

`SetupSpheresPipelines()` should be called exactly once, the appropriate pipeline will be built on demand from current knob values.

Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: Thanks @hysw  for confirming it. Updated.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp (hunk 14 above)
Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: Done.

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp (hunk 15 above)
Comment (elviscapiaq) [reply benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp]: Removed.

**Review by apazylbe — APPROVED** [review]
